service/policy/actions/actions.go

```diff
@@ -78,22 +79,125 @@ func NewRegistration(ns string, dbRegister serviceregistry.DBRegister) *servicer
+	a.logger.DebugContext(ctx, "listing actions")
+	rsp, err := a.dbClient.ListActions(ctx, req.Msg)
+	if err != nil {
+		return nil, db.StatusifyError(err, db.ErrTextListRetrievalFailed)
+	}
+	a.logger.DebugContext(ctx, "listed actions")
+	return connect.NewResponse(rsp), nil
 }
 
-func (a *ActionService) CreateAction(context.Context, *connect.Request[actions.CreateActionRequest]) (*connect.Response[actions.CreateActionResponse], error) {
-	return nil, connect.NewError(connect.CodeUnimplemented, errors.New("CreateAction is not implemented"))
+func (a *ActionService) CreateAction(ctx context.Context, req *connect.Request[actions.CreateActionRequest]) (*connect.Response[actions.CreateActionResponse], error) {
+	a.logger.DebugContext(ctx, "creating action", slog.String("name", req.Msg.GetName()))
+	auditParams := audit.PolicyEventParams{
+		ActionType: audit.ActionTypeCreate,
+		ObjectType: audit.ObjectTypeAction,
+	}
+	rsp := &actions.CreateActionResponse{}
+
+	err := a.dbClient.RunInTx(ctx, func(txClient *policydb.PolicyDBClient) error {
+		action, err := txClient.CreateAction(ctx, req.Msg)
+		if err != nil {
+			a.logger.Audit.PolicyCRUDFailure(ctx, auditParams)
+			return err
+		}
+
+		auditParams.ObjectID = action.GetId()
+		auditParams.Original = action
+		a.logger.Audit.PolicyCRUDSuccess(ctx, auditParams)
+
+		rsp.Action = action
+		return nil
+	})
+	if err != nil {
+		return nil, db.StatusifyError(err, db.ErrTextCreationFailed, slog.String("action", req.Msg.String()))
+	}
+	return connect.NewResponse(rsp), nil
 }
 
-func (a *ActionService) UpdateAction(context.Context, *connect.Request[actions.UpdateActionRequest]) (*connect.Response[actions.UpdateActionResponse], error) {
-	return nil, connect.NewError(connect.CodeUnimplemented, errors.New("UpdateAction is not implemented"))
+func (a *ActionService) UpdateAction(ctx context.Context, req *connect.Request[actions.UpdateActionRequest]) (*connect.Response[actions.UpdateActionResponse], error) {
+	actionID := req.Msg.GetId()
+	a.logger.DebugContext(ctx, "updating action", slog.String("id", actionID))
+	rsp := &actions.UpdateActionResponse{}
+
+	auditParams := audit.PolicyEventParams{
+		ActionType: audit.ActionTypeUpdate,
+		ObjectType: audit.ObjectTypeAction,
+		ObjectID:   actionID,
+	}
+
+	err := a.dbClient.RunInTx(ctx, func(txClient *policydb.PolicyDBClient) error {
+		original, err := txClient.GetAction(ctx, &actions.GetActionRequest{
+			Identifier: &actions.GetActionRequest_Id{
+				Id: actionID,
+			},
+		})
+		if err != nil {
+			a.logger.Audit.PolicyCRUDFailure(ctx, auditParams)
+			return db.StatusifyError(err, db.ErrTextGetRetrievalFailed, slog.String("id", actionID))
+		}
+
+		updated, err := txClient.UpdateAction(ctx, req.Msg)
+		if err != nil {
+			a.logger.Audit.PolicyCRUDFailure(ctx, auditParams)
+			return db.StatusifyError(err, db.ErrTextUpdateFailed, slog.String("id", actionID))
```

Comment:
Nit: I typically do `return db.StatusifyError(err, db.ErrTextUpdateFailed, slog.String("action", req.Msg.String()))` so that the log also includes the full request message just in case something in the message caused the error and you can hopefully see the problem immediately.  However, this is probably not consistent across the services so not a huge deal.